packages/model-viewer/src/features/scene-graph/model.ts

```diff
@@ -13,42 +13,275 @@
+          primitive.extensions['KHR_materials_variants']) {
+        const variantsExtension =
+            primitive.extensions['KHR_materials_variants'] as
+            KHRMaterialsVariants;
+        const extensions = threeGLTF.parser.json.extensions;
+        const variantNames = extensions['KHR_materials_variants'].variants;
+        // Provides definition now that we know there are variants to support.
+        this[$variantInfo] = new Map<string, Material>();
+        for (const mapping of variantsExtension.mappings!) {
+          // Maps variant indices to Materials.
+          this[$materials].set(mapping.material, mvMaterials[mapping.material]);
+          for (const variant of mapping.variants) {
+            const {name} = variantNames[variant];
+            this[$variantInfo].set(name, mvMaterials[mapping.material]);
+          }
+        }
+      }
+    }
+  }
+
+  get mesh() {
+    return this[$mesh];
+  }
+
+  async setActiveMaterial(material: number|Material):
+      Promise<ThreeMaterial|ThreeMaterial[]|null> {
+    if (material instanceof Material) {
+      this.mesh.material = await material[$ensureLoaded]();
+    } else {
+      const mvMaterial = this[$materials].get(material);
+      if (mvMaterial != null) {
+        this.mesh.material = await mvMaterial[$ensureLoaded]();
+      }
+    }
+    return this.mesh.material;
+  }
+
+  async enableVariant(name: string):
+      Promise<ThreeMaterial|ThreeMaterial[]|null> {
+    if (this[$variantInfo] != null) {
+      const material = this[$variantInfo].get(name);
+      if (material != null) {
+        return await this.setActiveMaterial(material);
+      }
+    }
+    return null;
+  }
+
+  async instantiateVariants() {
+    if (this[$variantInfo] == null) {
+      return;
+    }
+    for (const name of this[$variantInfo].keys()) {
+      if (this.mesh.userData.variantMaterials.get(name).material != null) {
+        continue;
+      }
+      const threeMaterial = await this.enableVariant(name);
+      if (threeMaterial != null) {
+        this.mesh.userData.variantMaterials.get(name).material = threeMaterial;
+      }
+    }
+  }
+
+  get variantInfo() {
+    return this[$variantInfo];
+  }
+}
 
 /**
  * A Model facades the top-level GLTF object returned by Three.js' GLTFLoader.
  * Currently, the model only bothers itself with the materials in the Three.js
  * scene graph.
  */
 export class Model implements ModelInterface {
-  private[$materials]: Array<Material> = [];
+  private[$materials] = new Array<Material>();
+  private[$hierarchy] = new Array<Node>();
+  private[$roots] = new Array<Node>();
+  private[$primitives] = new Array<PrimitiveNode>();
 
   constructor(
       correlatedSceneGraph: CorrelatedSceneGraph,
       onUpdate: () => void = () => {}) {
-    const {gltf, gltfElementMap} = correlatedSceneGraph;
+    const {gltf, threeGLTF, gltfElementMap} = correlatedSceneGraph;
+
+    for (let i = 0, size = gltf.materials!.length; i < size; ++i) {
```

Comment:
I believe this is the syntax you're looking for: `for (const [i, value] of myArray.entries())`

Reply:
Done.